### PR TITLE
CDAP-6196 adding join implementation to spark engine

### PR DESCRIPTION
**Pull request by albertshau**

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4432

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java [hunk 1]
```diff
@@ -18,50 +18,53 @@
...
 
   private static final String LOGICAL_START_TIME_FUNCTION_NAME = "logicalStartTime";
   private static final String SECURE_FUNCTION_NAME = "secure";
 
-  public DefaultMacroEvaluator(WorkflowToken workflowToken, Map<String, String> runtimeArguments,
+  public DefaultMacroEvaluator(@Nullable WorkflowToken workflowToken, Map<String, String> runtimeArguments,
                                long logicalStartTime) {
-    this.workflowToken = workflowToken;
-    this.runtimeArguments = runtimeArguments;
+    Map<String, String> arguments = new HashMap<>();
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java]: nit: should we rename this variable for clarity?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/SparkJoinerContext.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import co.cask.cdap.etl.batch.AbstractJoinerContext;
+import co.cask.cdap.etl.common.DatasetContextLookupProvider;
+
+/**
+ * Spark joiner context for preparing jobs.
+ */
+public class SparkJoinerContext extends AbstractJoinerContext {
+
+  protected SparkJoinerContext(String stageName, SparkClientContext scc) {
+    super(scc,
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/SparkJoinerContext.java]: nit: no need to have separate line for each parameter.

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinMergeFunction.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  private transient TrackedTransform<Tuple2<Object, List<JoinElement<Object>>>, Object> joinFunction;
+  private transient DefaultEmitter<Object> emitter;
+
+  public JoinMergeFunction(PluginFunctionContext pluginFunctionContext) {
+    this.pluginFunctionContext = pluginFunctionContext;
+  }
+
+  @Override
+  public Iterable<Object> call(Tuple2<Object, List<JoinElement<Object>>> input) throws Exception {
+    if (joinFunction == null) {
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinMergeFunction.java:46]: is there any reason not initializing the `joinFunction` in the constructor of this class?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinOnFunction.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  private transient DefaultEmitter<Tuple2<Object, Object>> emitter;
+
+  public JoinOnFunction(PluginFunctionContext pluginFunctionContext, String inputStageName) {
+    this.pluginFunctionContext = pluginFunctionContext;
+    this.inputStageName = inputStageName;
+  }
+
+  @Override
+  public Iterable<Tuple2<Object, Object>> call(Object input) throws Exception {
+    if (joinFunction == null) {
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinOnFunction.java:45]: same here.

**Comment by sagarkapare** [comment]
@albertshau minor nits and one question about variable initialization. Rest LGTM.

**Comment by CuriousVini** [comment]
LGTM

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java]: resolvedArguments? pipelineArguments?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinMergeFunction.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinMergeFunction.java:46]: yep. This function is run in a spark closure, which means Spark is responsible for constructing the object and populating the fields through deserialization. So all fields need to be serializable, or created through serializable fields.

I can add a comment to clarify.

**Comment by albertshau** [comment]
changes were only javadoc and a variable rename. merging since build was green.

**Pull request merged**